### PR TITLE
Feature/15429 associated services and xunit

### PR DESCRIPTION
**Pull request by MichaelRobinson-BJSS**

### CODE REVIEW

**Review by bjss-chris — CHANGES_REQUESTED** [review]
Looks alright overall. Quite a few comments but most of them are fairly small things.

There are a few classes where I've asked for a changed related to your PR that I've also asked for a small tech debt change (usually just making the class `sealed`).

There's still tech debt in a number of the classes this PR touches but we can fix those up separately.

File: src/NHSD.GPIT.BuyingCatalogue.Services/AssociatedServices/AssociatedServicesService.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using NHSD.GPIT.BuyingCatalogue.EntityFramework;
+using NHSD.GPIT.BuyingCatalogue.EntityFramework.Models.GPITBuyingCatalogue;
+using NHSD.GPIT.BuyingCatalogue.ServiceContracts.AssociatedServices;
+
+namespace NHSD.GPIT.BuyingCatalogue.Services.AssociatedServices
+{
+    public class AssociatedServicesService : IAssociatedServicesService
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.Services/AssociatedServices/AssociatedServicesService.cs]: Done

File: src/NHSD.GPIT.BuyingCatalogue.Services/AssociatedServices/AssociatedServicesService.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        private readonly GPITBuyingCatalogueDbContext dbContext;
+
+        public AssociatedServicesService(GPITBuyingCatalogueDbContext dbContext)
+        {
+            this.dbContext = dbContext ?? throw new ArgumentNullException(nameof(dbContext));
+        }
+
+        public async Task<List<CatalogueItem>> GetAssociatedServicesForSupplier(string supplierId)
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.Services/AssociatedServices/AssociatedServicesService.cs]: Done

File: src/NHSD.GPIT.BuyingCatalogue.Services/AssociatedServices/AssociatedServicesService.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+                .Include(x => x.Solution)
+                .ThenInclude(x => x.SolutionCapabilities)
+                .ThenInclude(x => x.Capability)
+                .Include(x => x.Supplier)
+                .Include(x => x.AssociatedService)
+                .Where(
+                    x => x.SupplierId == supplierId
+                        && x.CatalogueItemType == CatalogueItemType.AssociatedService
+                        && x.PublishedStatus == PublicationStatus.Published)
+                .OrderBy(x => x.Name)
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.Services/AssociatedServices/AssociatedServicesService.cs]: Done

File: src/NHSD.GPIT.BuyingCatalogue.Services/Session/OrderSessionService.cs [hunk 4]
```diff
@@ -102,8 +106,16 @@ public async Task<bool> InitialiseStateForEdit(string odsCode, CallOffId callOff
                 // CatalogueSolutionId = orderItem.CatalogueItem.ParentCatalogueItemId?.ToString(),
             };
 
-            var recipients = await odsService.GetServiceRecipientsByParentOdsCode(odsCode);
-            state.ServiceRecipients = recipients.Select(x => new OrderItemRecipientModel(x)).ToList();
+            if (state.CatalogueItemType == CatalogueItemType.AssociatedService)
+            {
+                var organisation = await organisationService.GetOrganisationByOdsCode(odsCode);
+                state.ServiceRecipients = new List<OrderItemRecipientModel> { new OrderItemRecipientModel { OdsCode = odsCode, Name = organisation.Name } };
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.Services/Session/OrderSessionService.cs]: Done

File: src/NHSD.GPIT.BuyingCatalogue.Services/Session/OrderSessionService.cs [hunk 5]
```diff
@@ -102,8 +106,16 @@ public async Task<bool> InitialiseStateForEdit(string odsCode, CallOffId callOff
...
+            }
+            else
+            {
+                var recipients = await odsService.GetServiceRecipientsByParentOdsCode(odsCode);
+                state.ServiceRecipients = recipients.Select(x => new OrderItemRecipientModel(x)).ToList();
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.Services/Session/OrderSessionService.cs]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Solutions/Controllers/DFOCVCController.cs [hunk 6]
```diff
@@ -10,12 +10,10 @@ namespace NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Solutions.Controllers
     [Area("Solutions")]
     public class DFOCVCController : Controller
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Solutions/Controllers/DFOCVCController.cs]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Solutions/Controllers/FuturesController.cs [hunk 7]
```diff
@@ -12,13 +12,11 @@ namespace NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Solutions.Controllers
     [Area("Solutions")]
     public class FuturesController : Controller
     {
-        private readonly ILogWrapper<FuturesController> logger;
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Solutions/Controllers/FuturesController.cs:15]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Solutions/Controllers/FuturesController.cs [hunk 8]
```diff
@@ -12,13 +12,11 @@ namespace NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Solutions.Controllers
     [Area("Solutions")]
     public class FuturesController : Controller
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Solutions/Controllers/FuturesController.cs]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Solutions/Controllers/VaccinationsController.cs [hunk 9]
```diff
@@ -1,17 +1,12 @@
...
     [Area("Solutions")]
     public class VaccinationsController : Controller
     {
-        private readonly ILogWrapper<VaccinationsController> logger;
-
-        public VaccinationsController(ILogWrapper<VaccinationsController> logger)
+        public VaccinationsController()
         {
-            this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
         }
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Solutions/Controllers/VaccinationsController.cs]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Solutions/Controllers/VaccinationsController.cs [hunk 10]
```diff
@@ -1,17 +1,12 @@
-﻿using System;
-using Microsoft.AspNetCore.Mvc;
-using NHSD.GPIT.BuyingCatalogue.Framework.Logging;
+﻿using Microsoft.AspNetCore.Mvc;
 
 namespace NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Solutions.Controllers
 {
     [Area("Solutions")]
     public class VaccinationsController : Controller
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Solutions/Controllers/VaccinationsController.cs]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/ActionFilters/SerilogMvcLoggingAttribute.cs [hunk 11]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.Extensions.DependencyInjection;
+using Serilog;
+
+namespace NHSD.GPIT.BuyingCatalogue.WebApp.ActionFilters
+{
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method)]
+    public class SerilogMvcLoggingAttribute : ActionFilterAttribute
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/ActionFilters/SerilogMvcLoggingAttribute.cs:9]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs [hunk 12]
```diff
@@ -17,20 +17,17 @@ namespace NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Admin.Controllers
     [Route("admin/organisations")]
     public class OrganisationsController : Controller
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Admin/Controllers/OrganisationsController.cs:18]: Done. Although AspNetCore.Routing is required

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs [hunk 13]
```diff
@@ -1,80 +1,247 @@
...
+
+            var organisation = await organisationService.GetOrganisationByOdsCode(odsCode);
+
+            var state = new CreateOrderItemModel
+            {
+                IsNewOrder = true,
+                CommencementDate = order.CommencementDate,
+                SupplierId = order.SupplierId,
+                CatalogueItemType = CatalogueItemType.AssociatedService,
+                ServiceRecipients = new List<OrderItemRecipientModel> { new OrderItemRecipientModel { OdsCode = odsCode, Name = organisation.Name } },
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs:74]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs [hunk 14]
```diff
@@ -1,80 +1,247 @@
...
+            }
+
+            var existingOrder = await orderItemService.GetOrderItem(callOffId, model.SelectedSolutionId.Value);
+
+            if (existingOrder != null)
+            {
+                return RedirectToAction(
+                    actionName: nameof(EditAssociatedService),
+                    controllerName: typeof(AssociatedServicesController).ControllerName(),
+                    routeValues: new { odsCode, callOffId, id = existingOrder.CatalogueItemId });
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs:105]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs [hunk 15]
```diff
@@ -1,80 +1,247 @@
...
+            }
+
+            state.CatalogueItemId = model.SelectedSolutionId;
+            var solution = await solutionsService.GetSolution(state.CatalogueItemId.Value);
+            state.CatalogueItemName = solution.Name;
+            state.CatalogueSolutionId = model.SelectedSolutionId;
+
+            orderSessionService.SetOrderStateToSession(state);
+
+            var prices = solution.CataloguePrices.Where(x => x.CataloguePriceType == CataloguePriceType.Flat).ToList();
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs:115]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs [hunk 16]
```diff
@@ -1,80 +1,247 @@
...
+
+            if (prices.Count == 1)
+            {
+                orderSessionService.SetPrice(prices.Single());
+
+                return RedirectToAction(
+                    actionName: nameof(EditAssociatedService),
+                    controllerName: typeof(AssociatedServicesController).ControllerName(),
+                    routeValues: new { odsCode, callOffId, id = state.CatalogueItemId });
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs:124]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs [hunk 17]
```diff
@@ -1,80 +1,247 @@
...
+            }
+
+            return RedirectToAction(
+                actionName: nameof(SelectAssociatedServicePrice),
+                controllerName: typeof(AssociatedServicesController).ControllerName(),
+                routeValues: new { odsCode, callOffId });
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs:130]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs [hunk 18]
```diff
@@ -1,80 +1,247 @@
...
-        public IActionResult NewOrderItem(string odsCode, string callOffId)
+        [HttpGet("select/associated-service/price")]
+        public async Task<IActionResult> SelectAssociatedServicePrice(string odsCode, CallOffId callOffId)
         {
-            return View(new NewAssociatedServiceOrderItemModel());
+            var state = orderSessionService.GetOrderStateFromSession();
+
+            var solution = await solutionsService.GetSolution(state.CatalogueItemId.Value);
+
+            var prices = solution.CataloguePrices.Where(x => x.CataloguePriceType == CataloguePriceType.Flat).ToList();
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs:140]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs [hunk 19]
```diff
@@ -1,80 +1,247 @@
...
+        public async Task<IActionResult> SelectAssociatedServicePrice(string odsCode, CallOffId callOffId, SelectAssociatedServicePriceModel model)
         {
-            return Redirect($"/order/organisation/{odsCode}/order/{callOffId}/associated-services");
+            var state = orderSessionService.GetOrderStateFromSession();
+
+            var solution = await solutionsService.GetSolution(state.CatalogueItemId.Value);
+
+            if (!ModelState.IsValid)
+            {
+                model.SetPrices(solution.CataloguePrices.Where(x => x.CataloguePriceType == CataloguePriceType.Flat).ToList());
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs:154]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs [hunk 20]
```diff
@@ -1,80 +1,247 @@
...
+                return View(model);
+            }
+
+            var cataloguePrice = solution.CataloguePrices.Single(x => x.CataloguePriceId == model.SelectedPrice);
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs:158]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs [hunk 21]
```diff
@@ -1,80 +1,247 @@
...
+
+            orderSessionService.SetPrice(cataloguePrice);
+
+            return RedirectToAction(
+                actionName: nameof(EditAssociatedService),
+                controllerName: typeof(AssociatedServicesController).ControllerName(),
+                routeValues: new { odsCode, callOffId, id = state.CatalogueItemId });
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs:165]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs [hunk 22]
```diff
@@ -1,80 +1,247 @@
...
+
+            // TODO - handle errors
+            var result = await orderItemService.Create(callOffId, state);
+
+            sessionService.ClearSession();
+
+            return RedirectToAction(
+                actionName: nameof(Index),
+                controllerName: typeof(AssociatedServicesController).ControllerName(),
+                routeValues: new { odsCode, callOffId });
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs:225]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs [hunk 23]
```diff
@@ -1,80 +1,247 @@
...
+            await orderItemService.DeleteOrderItem(callOffId, id);
 
-        [HttpGet("delete/{id}/confirmation/{serviceName}/continue")]
-        public IActionResult DeleteContinue(string odsCode, string callOffId, string id, string serviceName)
-        {
-            return View(new DeleteContinueModel());
+            return RedirectToAction(
+                actionName: nameof(Index),
+                controllerName: typeof(AssociatedServicesController).ControllerName(),
+                routeValues: new { odsCode, callOffId });
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/AssociatedServicesController.cs:244]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/DashboardController.cs [hunk 24]
```diff
@@ -113,7 +97,7 @@ public async Task<IActionResult> NewOrderDescription(string odsCode, NewOrderDes
             return RedirectToAction(
                 actionName: nameof(OrderController.Order),
                 controllerName: typeof(OrderController).ControllerName(),
-                routeValues: new { odsCode = odsCode, callOffId = order.CallOffId });
+                routeValues: new { odsCode, callOffId = order.CallOffId });
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Controllers/DashboardController.cs:100]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Models/AssociatedServices/DeleteAssociatedServiceModel.cs [hunk 25]
```diff
@@ -1,12 +1,31 @@
...
+namespace NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Order.Models.AssociatedServices
 {
     public class DeleteAssociatedServiceModel : OrderingBaseModel
     {
         public DeleteAssociatedServiceModel()
         {
-            BackLink = "/order/organisation/03F/order/C010000-01/associated-services/10000-S-003"; // TODO
+        }
+
+        public DeleteAssociatedServiceModel(string odsCode, CallOffId callOffId, string orderItemId, string solutionName, string orderDescription)
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Models/AssociatedServices/DeleteAssociatedServiceModel.cs:11]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Models/AssociatedServices/DeleteAssociatedServiceModel.cs [hunk 26]
```diff
@@ -1,12 +1,31 @@
...
+            OdsCode = odsCode;
+            CallOffId = callOffId;
+            OrderItemId = orderItemId;
+            SolutionName = solutionName;
+            OrderDescription = orderDescription;
         }
+
+        public CallOffId CallOffId { get; set; }
+
+        public string OrderItemId { get; set; }
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Models/AssociatedServices/DeleteAssociatedServiceModel.cs:25]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Models/AssociatedServices/DeleteAssociatedServiceModel.cs [hunk 27]
```diff
@@ -1,12 +1,31 @@
-﻿namespace NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Order.Models.AssociatedServices
+﻿using NHSD.GPIT.BuyingCatalogue.EntityFramework.Ordering.Models;
+
+namespace NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Order.Models.AssociatedServices
 {
     public class DeleteAssociatedServiceModel : OrderingBaseModel
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Models/AssociatedServices/DeleteAssociatedServiceModel.cs:5]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Models/AssociatedServices/EditAssociatedServiceModel.cs [hunk 28]
```diff
@@ -1,12 +1,50 @@
-﻿namespace NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Order.Models.AssociatedServices
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using NHSD.GPIT.BuyingCatalogue.EntityFramework.Models.GPITBuyingCatalogue;
+using NHSD.GPIT.BuyingCatalogue.EntityFramework.Ordering.Models;
+using NHSD.GPIT.BuyingCatalogue.Framework.Constants;
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Models/AssociatedServices/EditAssociatedServiceModel.cs:6]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Models/AssociatedServices/EditAssociatedServiceModel.cs [hunk 29]
```diff
@@ -1,12 +1,50 @@
...
+using NHSD.GPIT.BuyingCatalogue.ServiceContracts.Models;
+
+namespace NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Order.Models.AssociatedServices
 {
     public class EditAssociatedServiceModel : OrderingBaseModel
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Models/AssociatedServices/EditAssociatedServiceModel.cs:11]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Models/AssociatedServices/NoAssociatedServicesFoundModel.cs [hunk 30]
```diff
@@ -0,0 +1,14 @@
+﻿using NHSD.GPIT.BuyingCatalogue.EntityFramework.Ordering.Models;
+
+namespace NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Order.Models.AssociatedServices
+{
+    public class NoAssociatedServicesFoundModel : OrderingBaseModel
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Models/AssociatedServices/NoAssociatedServicesFoundModel.cs:5]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Models/AssociatedServices/SelectAssociatedServicePriceModel.cs [hunk 31]
```diff
@@ -0,0 +1,41 @@
+﻿using System.Collections.Generic;
+using NHSD.GPIT.BuyingCatalogue.EntityFramework.Models.GPITBuyingCatalogue;
+using NHSD.GPIT.BuyingCatalogue.EntityFramework.Ordering.Models;
+using NHSD.GPIT.BuyingCatalogue.Framework.Constants;
+
+namespace NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Order.Models.AssociatedServices
+{
+    public class SelectAssociatedServicePriceModel : OrderingBaseModel
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Models/AssociatedServices/SelectAssociatedServicePriceModel.cs:8]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Views/AssociatedServices/DeleteAssociatedService.cshtml [hunk 32]
```diff
@@ -3,41 +3,30 @@
...
-                            Cancel delete
-                        </a>
-                    </div>
-                </div>
+                    <nhs-submit-button text="Confirm Delete" />
+                    <vc:nhs-backlink text="Cancel delete"
+                                     url="@Url.Action(
+                                            action: "EditAssociatedService",
+                                            controller: "AssociatedServices",
+                                            values: new { odsCode = Model.OdsCode,
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Views/AssociatedServices/DeleteAssociatedService.cshtml:26]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Views/AssociatedServices/EditAssociatedService.cshtml [hunk 33]
```diff
@@ -5,143 +5,94 @@
...
+
+                <nhs-submit-button text="Save" /><br/>
+
+                @if (!Model.OrderItem.IsNewOrder)
+                {
+                    <vc:nhs-delete-button text="Delete Associated Service"
+                                          url="@Url.Action(
+                                          action: "DeleteAssociatedService",
+                                          controller: "AssociatedServices",
+                                          values: new { odsCode = Model.OdsCode,
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Views/AssociatedServices/EditAssociatedService.cshtml:83]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Views/AssociatedServices/Index.cshtml [hunk 34]
```diff
@@ -8,66 +8,52 @@
...
+
+            <nhs-fieldset-form-container asp-for="@Model"
+                                         label-text="Order description"
+                                         label-hint="@Model.OrderDescription"
+                                         size="Large">
+
+                <vc:nhs-secondary-button url="@Url.Action(
+                                                  action: "SelectAssociatedService",
+                                                  controller:"AssociatedServices",
+                                                  values: new {odsCode = Model.OdsCode, CallOffId = Model.CallOffId })"
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Views/AssociatedServices/Index.cshtml:22]: Done - Model.OdsCode works

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Views/AssociatedServices/Index.cshtml [hunk 35]
```diff
@@ -8,66 +8,52 @@
...
+                            <nhs-table-cell>
+                                <a asp-action="EditAssociatedService"
+                                   asp-controller="AssociatedServices"
+                                   asp-route-id ="@item.CatalogueItemId"
+                                   asp-route-odsCode="@Model.OdsCode"
+                                   asp-route-callOffId="@Model.CallOffId">
+                                @item.CatalogueItem.Name</a>
+                            </nhs-table-cell>
+                            <nhs-table-cell>
+                                @item.CataloguePrice.PricingUnit.Description @item?.CataloguePrice.TimeUnit
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Views/AssociatedServices/Index.cshtml:40]: Done

File: src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Views/AssociatedServices/NoAssociatedServicesFound.cshtml [hunk 36]
```diff
@@ -0,0 +1,10 @@
+﻿@model NHSD.GPIT.BuyingCatalogue.WebApp.Areas.Order.Models.AssociatedServices.NoAssociatedServicesFoundModel;
+@{
+    ViewData["Title"] = Model.Title;
+}
+
+<div class="nhsuk-width-container nhsuk-u-margin-top-4 nhsuk-u-margin-bottom-4 bc-c-error-container nhsuk-u-padding-4">
```
[the remark this answers is not in the document]
Comment (MichaelRobinson-BJSS) [reply src/NHSD.GPIT.BuyingCatalogue.WebApp/Areas/Order/Views/AssociatedServices/NoAssociatedServicesFound.cshtml:6]: Done

**Review by bjss-chris — APPROVED** [review]

**Pull request merged**